### PR TITLE
Bf16 routed moe

### PR DESCRIPTION
**Pull request by IwakuraRein**
## 📌 Description

Add `trtllm_bf16_routed_moe` api

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

`pytest tests/moe/test_trtllm_gen_routed_fused_moe.py::test_trtllm_gen_bf16_routed_fused_moe`

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added support for pre-computed routing in MoE operations, enabling flexible routing input strategies.
  * New routed MoE APIs now available: BF16 and FP8 variants support pre-packed top-k routing information.
  * Introduced dual-path mechanism allowing MoE operations to accept either routing logits or pre-computed routing data.

### CODE REVIEW

File: flashinfer/fused_moe/core.py [hunk 1]
```diff
@@ -1340,12 +1353,22 @@ def trtllm_bf16_moe_op(
...
+        else:
+            # When routing_logits is provided, we either have topk_ids/expert_weights,
+            # packed into a single tensor as topk_id
+            # or have them individually as topk_ids and expert_weights respectively
+            topk_ids = topk_ids
+            expert_weights = (
+                expert_weights
+                if expert_weights is not None
+                else torch.empty(0, dtype=torch.bfloat16, device=hidden_states.device)
+            )
```
[the remark this answers is not in the document]
Comment (IwakuraRein) [reply flashinfer/fused_moe/core.py:1371]: If `topk_ids` is not None and `expert_weights` is None, the expert weights are encoded in the `topk_ids`. We need to refactor all the routed moe apis in the near future to make it more clear.

File: flashinfer/fused_moe/core.py [hunk 2]
```diff
@@ -1725,12 +1775,24 @@ def trtllm_fp8_block_scale_moe_op(
...
+        if do_finalize:
+            return [output]
+        else:
+            return [
+                torch.from_dlpack(intermediate_output[0]),
+                (
+                    torch.from_dlpack(intermediate_output[1])
+                    if routing_logits is not None or expert_weights.numel() == 0
+                    else expert_weights
+                ),
```
[the remark this answers is not in the document]
Comment (IwakuraRein) [reply flashinfer/fused_moe/core.py:1793]: if `routing_logits` is not None, the expert weights are computed by the flashinfer routing kernels, so we return the kernel output. Otherwise, the topk ids and the expert weights are provided by the user through the routed moe APIs. 

In that case, I don't think there is need for returning the expert weights. In the routed moe APIs comments, it marked the second return as undefined.

File: flashinfer/fused_moe/core.py [hunk 3]
```diff
@@ -1312,8 +1314,10 @@ def refine_tuning_config(cls, tune_max_num_tokens: int):
         mutates_args=(""),
     )
     def trtllm_bf16_moe_op(
-        routing_logits: torch.Tensor,
+        routing_logits: Optional[torch.Tensor],
         routing_bias: Optional[torch.Tensor],
+        topk_ids: Optional[torch.Tensor],
```
Comment (aleozlx) [inline flashinfer/fused_moe/core.py:1319]: for compatibility we have to add these `topk_ids` and `expert_weights` at the end with default=None, despite they are gonna be a bit different than other data types

**Comment by aleozlx** [comment]
~tests are good. need to resolve one api compatibility issue as above. otherwise lgtm~

File: flashinfer/fused_moe/core.py (hunk 3 above)
Comment (IwakuraRein) [reply flashinfer/fused_moe/core.py:1319]: `trtllm_bf16_moe_op` was not exposed to users. Instead of put the `topk_ids` and `expert_weights` at the end, I reckon it's better to put them at the top to align with other moe_ops, like `trtllm_fp8_block_scale_moe_op`

**Review by aleozlx — APPROVED** [review]
lgtm

**Comment by aleozlx** [comment]
this is ready for merging
